File: ObjC/Restaurant/Common/KeyboardListener.m
//
//  KeyboardListener.m
//  Restaurant
//
//  Created by AppsFoundation on 2/19/15.
//  Copyright (c) 2015 AppsFoundation. All rights reserved.
//

#import "KeyboardListener.h"

@interface KeyboardListener ()

@property NSInteger contraintCanstant;

@end

@implementation KeyboardListener

- (id)initWithScrollView:(UIScrollView *)scrollView andConstraint:(NSLayoutConstraint *)contraint {
    if (!self) {
        self = [self init];
    }
    
    self.scrollView = scrollView;
    self.contraint = contraint;
    self.contraintCanstant = self.contraint.constant;
    
    //keyboard notifications to change tableview frame
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillShow:) name:UIKeyboardWillChangeFrameNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillHide:) name:UIKeyboardWillHideNotification object:nil];
    
    return self;
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

#pragma mark - Keyboard notifications

- (void)keyboardWillShow:(NSNotification *)notification {
    NSDictionary *info = [notification userInfo];
    NSValue *kbFrame = [info objectForKey:UIKeyboardFrameEndUserInfoKey];
    NSTimeInterval animationDuration = [[info objectForKey:UIKeyboardAnimationDurationUserInfoKey] doubleValue];
    CGRect keyboardFrame = [kbFrame CGRectValue];
    
    CGFloat height = keyboardFrame.size.height;
    self.contraint.constant = height;
    
    [UIView animateWithDuration:animationDuration animations:^{
        [self.scrollView.superview layoutIfNeeded];
    }];
}

- (void)keyboardWillHide:(NSNotification *)notification {
    NSDictionary *info = [notification userInfo];
    NSTimeInterval animationDuration = [[info objectForKey:UIKeyboardAnimationDurationUserInfoKey] doubleValue];
    
    self.contraint.constant = self.contraintCanstant;
    [UIView animateWithDuration:animationDuration animations:^{
        [self.scrollView.superview layoutIfNeeded];
    }];
}

@end

File: Swift/Restaurant/Common/KeyboardListener.swift
//
//  KeyboardListener.swift
//  Restaurant
//
//  Created by AppsFoundation on 8/27/15.
//  Copyright © 2015 AppsFoundation. All rights reserved.
//

import UIKit

class KeyboardListener: NSObject {
    
    var scrollView: UIScrollView
    var constraint: NSLayoutConstraint
    
    private var constraintConstant: CGFloat
    
    init(scrollView: UIScrollView, constraint: NSLayoutConstraint) {
        self.scrollView = scrollView
        self.constraint = constraint
        constraintConstant = self.constraint.constant
        
        super.init()
        
        //keyboard notifications to change tableview frame
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillShow), name: NSNotification.Name.UIKeyboardWillChangeFrame, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillHide), name: NSNotification.Name.UIKeyboardWillHide, object: nil)
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    // MARK: - Keyboard notifications
    func keyboardWillShow(notification: NSNotification) {
        if let info = notification.userInfo, let kbFrame: NSValue = info[UIKeyboardFrameEndUserInfoKey] as? NSValue, let animationDuration: TimeInterval = info[UIKeyboardAnimationDurationUserInfoKey] as? TimeInterval {
            
            let keyboardFrame = kbFrame.cgRectValue
            constraint.constant = keyboardFrame.size.height
            
            UIView.animate(withDuration: animationDuration, animations: { () -> Void in
                if let superview = self.scrollView.superview {
                    superview.layoutIfNeeded()
                }
            })
        }
    }
    
    func keyboardWillHide(notification: NSNotification) {
        if let info = notification.userInfo,
            
            let animationDuration: TimeInterval = info[UIKeyboardAnimationDurationUserInfoKey] as? TimeInterval {
            constraint.constant = constraintConstant
            
            UIView.animate(withDuration: animationDuration, animations: { () -> Void in
                if let superview = self.scrollView.superview {
                    superview.layoutIfNeeded()
                }
            })
        }
    }
}





